File: php/features.php
<?php
declare(strict_types=1);

// Commonroom SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class CommonroomFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new CommonroomBaseFeature();
            case "debug":
                return new CommonroomDebugFeature();
            case "idempotency":
                return new CommonroomIdempotencyFeature();
            case "metrics":
                return new CommonroomMetricsFeature();
            case "paging":
                return new CommonroomPagingFeature();
            case "ratelimit":
                return new CommonroomRatelimitFeature();
            case "retry":
                return new CommonroomRetryFeature();
            case "test":
                return new CommonroomTestFeature();
            case "timeout":
                return new CommonroomTimeoutFeature();
            default:
                return new CommonroomBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/commonroom_sdk/features.py
# Commonroom SDK feature factory

from commonroom_sdk.feature.base_feature import CommonroomBaseFeature
from commonroom_sdk.feature.debug_feature import CommonroomDebugFeature
from commonroom_sdk.feature.idempotency_feature import CommonroomIdempotencyFeature
from commonroom_sdk.feature.metrics_feature import CommonroomMetricsFeature
from commonroom_sdk.feature.paging_feature import CommonroomPagingFeature
from commonroom_sdk.feature.ratelimit_feature import CommonroomRatelimitFeature
from commonroom_sdk.feature.retry_feature import CommonroomRetryFeature
from commonroom_sdk.feature.test_feature import CommonroomTestFeature
from commonroom_sdk.feature.timeout_feature import CommonroomTimeoutFeature


_FEATURES = {
    "base": lambda: CommonroomBaseFeature(),
    "debug": lambda: CommonroomDebugFeature(),
    "idempotency": lambda: CommonroomIdempotencyFeature(),
    "metrics": lambda: CommonroomMetricsFeature(),
    "paging": lambda: CommonroomPagingFeature(),
    "ratelimit": lambda: CommonroomRatelimitFeature(),
    "retry": lambda: CommonroomRetryFeature(),
    "test": lambda: CommonroomTestFeature(),
    "timeout": lambda: CommonroomTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
